File: usaco/training/section_5.4/telecow.py
"""
ID: huangch7
LANG: PYTHON3
TASK: telecow
"""
# import math
# import copy
task_name = "telecow"
INF = 100000000


def get_list_from_line(dtype=int):
    return [dtype(x) for x in fin.readline().strip().split()]


def get_strs_from_line():
    return fin.readline().strip().split()



fin = open(f'{task_name}.in', 'r')
fout = open(f'{task_name}.out', 'w')

N, M, start, target = get_list_from_line()
adj = {i: {} for i in range(1, N + 1)}
adj_init = {i: {} for i in range(1, N + 1)}
for _ in range(M):
    s, e = get_list_from_line()
    adj[s][e] = 1
    adj[e][s] = 1
    adj_init[s][e] = 1
    adj_init[e][s] = 1





def max_flow(start, target, adj):

    def bfs(start, target, adj):
        level = {k: 0 for k in adj}
        queue = [start]
        level[start] = 1
        while len(queue):
            cur = queue.pop(0)
            for x in adj[cur]:
                if level[x]:
                    continue
                if adj[cur][x] <= 0:
                    continue
                level[x] = level[cur] + 1
                queue.append(x)
        return level[target] > 0, level

    def dfs(start, target, adj, cap, level):
        left = cap
        if start == target:
            return cap
        can_del = []
        for x in adj[start]:
            if level[x] == level[start] + 1 and adj[start][x] > 0:
                cur_flow = dfs(x, target, adj, min(left, adj[start][x]), level)
                adj[start][x] -= cur_flow
                if adj[start][x] == 0:
                    can_del.append(x)
                if start not in adj[x]:
                    adj[x][start] = 0
                adj[x][start] += cur_flow
                left -= cur_flow
#        for x in can_del:
#            del adj[start][x]
        return cap - left
                

    total = 0
    flow = {}
    for i in adj:
        flow[i] = {}
        for j in adj:
            flow[i][j] = 0

    while True:
        bfs_ok, level = bfs(start, target, adj)
        if not bfs_ok:
            break
#         print('level', level)
        total += dfs(start, target, adj, INF, level)
#         print(total, adj)

    return total
    
# print('debug', max_flow(1, 6, {1: {2:10, 4: 8}, 2:{3:10, 5:6}, 3:{6:10},
#                               4:{3:8}, 5:{6:6}, 6:{}}))

# print('debug', max_flow(start, target, adj))
        
    


# 点割变边割


def get_new_adj(old_adj, remove_node):
    """
    i -> j
    
    i1 -> i2
    j1 -> j2

    
    """
    adj = {}
    for i in old_adj:
        adj[i + 1000] = {}
        adj[i + 2000] = {}
    for i in old_adj:
        if i in remove_node:
            continue
        adj[i + 1000][i + 2000] = 1
        for j in old_adj[i]:
            if j in remove_node:
                continue
            adj[i + 2000][j + 1000] = INF
            adj[j + 2000][i + 1000] = INF
            
            
    return adj

cnt = 0
ans = []

total = max_flow(start + 2000, target + 1000, get_new_adj(adj_init, remove_node=[]))
# 1 -- 3 -\ / 6 - 2   
#.  +- 4 - 9 -  7 --+
#.  +- 5 -/ \- 8 --+


# print(total)
for i in range(1, N + 1):
    if i == start or i == target:
        continue
    
#     adj = copy.deepcopy(adj_init)
#    print(ans)
#    print(adj_init)
#    print(get_new_adj(adj_init, remove_node=ans))
    
#    cur = max_flow(start, target, adj, remove_node=ans + [i])
    cur = max_flow(start + 2000, target + 1000, get_new_adj(adj_init, remove_node=ans + [i]))
    # ans 里包含了已经切断的节点，每切断一个，当前最大流就少一个。
    # 现在寻找能让最大流再减少的节点。
#     print(i, cur)

    if cur == total - len(ans) - 1:
        ans.append(i)
        cnt += 1
    if cnt == total:
        break
    
# print(ans)

fout.write(f"{total}\n")
line = " ".join([str(x) for x in ans])
fout.write(f"{line}\n")

fout.close()

File: usaco/training/section_5.4/telecow.cpp
/*
ID: huangch7
LANG: C++
TASK: telecow

 */
#include <iostream>
#include <fstream>
#include <map>
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

const int INF = 100000000;

map<int, int>  bfs(int start, int target, map<int, map<int, int> >& adj){
  map<int, int> level;
  for (auto kv: adj) level[kv.first] = 0;
  level[start] = 1;
  queue<int> q;
  q.push(start);
  while (!q.empty()){
    int cur = q.front();
    q.pop();
    for (auto kv: adj[cur]){
      int x = kv.first;
      int w = kv.second;
      if (level[x] == 0 && w > 0){
        level[x] = level[cur] + 1;
        q.push(x);
      }
      
    }
    
  }
  return level;
  
}

int dfs(int start, int target, map<int, map<int, int> >& adj, int cap, map<int, int>& level){
  int left = cap;
  if (start == target) return cap;
  for (auto kv: adj[start]){
    int x = kv.first;
    int f = kv.second;
    if ((level[x] == level[start] + 1) && f > 0){
      int cur_flow = dfs(x, target, adj, min(left, f), level);
      adj[start][x] -= cur_flow;
      if (adj[x].find(start) == adj[x].end()) adj[x][start] = 0;
      adj[x][start] += cur_flow;
      left -= cur_flow;
                                                
    }
  }
  return cap - left;
}

int max_flow(int start, int target, map<int, map<int, int> >& adj){
  int total = 0;
  while (true){
    map<int, int> level = bfs(start, target, adj);
    if (level[target] == 0) break;
    total += dfs(start, target, adj, INF, level);
  }
  return total;
}


int max_flow_2(int start, int target, map<int, map<int, int> >& adj, const vector<int>& remove_node) {
    int total = 0;
    while (true) {
        map<int, int> flow, prev, visited;
        for (auto kv : adj) {
            flow[kv.first] = -1;
            prev[kv.first] = -1;
            visited[kv.first] = 0;
        }
        flow[start] = INF;

        for (int x : remove_node) {
            visited[x] = 1;
        }

        while (true) {
            int best = 0;
            int best_pos = 0;

            for (auto& kv : adj) {
              int i = kv.first;
                if (visited[i]) continue;
                if (flow[i] == -1) continue;
                if (best < flow[i]) {
                    best = flow[i];
                    best_pos = i;
                }
            }

            if (best_pos == 0 || best_pos == target) break;

            visited[best_pos] = 1;
            for (auto& kv : adj[best_pos]) {
              int j = kv.first;
              int val = kv.second;
                int cur_flow = min(val, flow[best_pos]);
                if (flow[j] == -1 || flow[j] < cur_flow) {
                    flow[j] = cur_flow;
                    prev[j] = best_pos;
                }
            }
        }

        if (flow[target] == -1) break;

        int cur = target;
        while (cur != start) {
            adj[prev[cur]][cur]--;
            if (adj[cur].find(prev[cur]) == adj[cur].end()) {
                adj[cur][prev[cur]] = 0;
            }
            adj[cur][prev[cur]]++;
            cur = prev[cur];
        }
        total += flow[target];
    }
    return total;
}


std::map<int, std::map<int, int>> get_new_adj(const std::map<int, std::map<int, int>>& old_adj, const std::vector<int>& remove_node) {
    std::map<int, std::map<int, int>> adj;

    for (const auto& i : old_adj) {
        if (std::find(remove_node.begin(), remove_node.end(), i.first) != remove_node.end()) {
            continue;
        }

        int i_1000 = i.first + 1000;
        int i_2000 = i.first + 2000;

        adj[i_1000] = {};
        adj[i_2000] = {};

        adj[i_1000][i_2000] = 1;

        for (const auto& j : i.second) {
            if (std::find(remove_node.begin(), remove_node.end(), j.first) != remove_node.end()) {
                continue;
            }

            int j_1000 = j.first + 1000;
            int j_2000 = j.first + 2000;

            adj[i_2000][j_1000] = INF;
            adj[j_2000][i_1000] = INF;
        }
    }

    return adj;
}

int main() {
    ifstream fin("telecow.in");
    ofstream fout("telecow.out");

    int N, M, start, target;
    fin >> N >> M >> start >> target;

    map<int, map<int, int>> adj, adj_init, adj_new;

    for (int i = 1; i <= N; ++i) {
        adj[i] = {};
        adj_init[i] = {};
    }

    for (int i = 0; i < M; ++i) {
        int s, e;
        fin >> s >> e;
        adj[s][e] = 1;
        adj[e][s] = 1;
        adj_init[s][e] = 1;
        adj_init[e][s] = 1;
    }
    adj_new = get_new_adj(adj_init, {});
    int total = max_flow(start + 2000, target + 1000, adj_new);

    int cnt = 0;
    vector<int> ans;

    for (int i = 1; i <= N; ++i) {
        if (i == start || i == target) continue;

        adj = adj_init;
        adj_new = get_new_adj(adj_init, {});
        vector<int> remove_node;
        for (auto x: ans) remove_node.push_back(x);
        remove_node.push_back(i);
        adj_new = get_new_adj(adj_init, remove_node);

        int cur = max_flow(start + 2000, target + 1000, adj_new);
        
        if (cur == total - ans.size() - 1) {
            ans.push_back(i);
            cnt++;
        }

        if (cnt == total) break;
    }

    fout << total << endl;

    string line;
    for (int i = 0; i < ans.size(); ++i) {
        line += to_string(ans[i]);
        if (i != ans.size() - 1) line += " ";
    }

    fout << line << endl;

    fin.close();
    fout.close();

    return 0;
}
